File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
class Solution {
public:
    // Making an array to represent directions
    // Specifically, up, down, left, and right
    vector<int> dx = {-1, 1, 0, 0};
    vector<int> dy = {0, 0, -1, 1};
    
    void DFS(vector<vector<char>>& grid, int row, int col) {
        // Bounds checking so we stop
        if (row < 0 || row >= grid.size() || col < 0 || col >= grid[0].size()) {
            return;
        } else if (grid[row][col] == '0') {
            // Stop since it's a water tile as well
            // This will be our way of marking visited tiles too
            return;
        }

        // Mark the current cell as visited by turning
        // it into a water tile
        grid[row][col] = '0';

        // Explore all adjacent cells
        for (int i = 0; i < dx.size(); i++) {
            DFS(grid, row + dx[i], col + dy[i]);
        }
    }
    
    int numIslands(vector<vector<char>>& grid) {
        // Tracking the total number of islands
        int totalIslands = 0;

        // Traversing the entire board
        for (int row = 0; row < grid.size(); row++) {
            for (int col = 0; col < grid[0].size(); col++) {
                // Check if the current cell is an island
                if (grid[row][col] == '1') {
                    // If it is, increment total islands and explore
                    totalIslands++;
                    DFS(grid, row, col);
                }
            }
        }

        // Return the total number of islands found
        return totalIslands;
    }
};

File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        # Will store the max area we find
        maxArea = 0

        # Defining all possible directions
        directions = [
            (-1, 0), # up
            (1, 0),  # down
            (0, -1), # left
            (0, 1),  # right
        ]

        # Create our DFS function
        def DFS(row, col):
            # Don't contribute if we are out of bounds
            if row >= len(grid) or row < 0 or col >= len(grid[0]) or col < 0:
                return 0

            # Don't contribute if it's a water tile
            if grid[row][col] == 0:
                return 0

            # Mark it as visited (make it a water tile)
            grid[row][col] = 0

            # We know we have one island tile, so
            # we automatically have an area of 1
            area = 1

            # Explore all possible directions
            for dr, dc in directions:
                # Get our new row and column positions
                newRow = row + dr
                newCol = col + dc

                # Add any adjacent island tiles
                area += DFS(newRow, newCol)

            # Return the ultimate total area we find
            return area

        # Traverse through the board
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == 1:
                    maxArea = max(maxArea, DFS(row, col))

        # Return the max area we have found
        return maxArea